### DISCUSSION TITLE
Inject TCP packets

### DISCUSSION THREAD

**Discussion by nathan-baggs** [mitmproxy/mitmproxy#6855] [Q&A]
Just wondering if it's possible to inject additional TCP packets through the addons API.

I've been doing this:
```python
def tcp_message(self, flow):
    flow.messages[-1].content += b"..."
```

But (I'm assuming) that gets forwarded as a single packet. What I want to achieve is something like this:

```python
def tcp_message(self, flow):
    flow.messages.append(copy.deepcopy(flow.messages[-1])
    flow.messages[-1].content = b"..."
```

But that doesn't seem to work for me, is such a thing even possible?

To provide some more context I want to modify a packet with data that will exceed the MTU so would like to manually split it up

**Comment by Prinzhorn** [comment]
There's an [`inject.tcp`](https://github.com/mitmproxy/mitmproxy/blob/0503fbb63a7d27fbaaa04a228203a82d6732dea4/mitmproxy/addons/proxyserver.py#L324) command.

You should be able to do sth. like

```py
ctx.master.commands.call("inject.tcp", flow, True, b"...")
```

**Reply by nathan-baggs** [reply]
Thanks this looks like what I want, however if I call this from within `tcp_message` I get:
```
Traceback (most recent call last):
  File "mitmproxy\addonmanager.py", line 47, in safecall
  File "mitmproxy\addonmanager.py", line 292, in trigger_event
  File "mitmproxy\addonmanager.py", line 269, in invoke_addon
  File "mitmproxy\addons\dumper.py", line 375, in tcp_message
  File "mitmproxy\addons\dumper.py", line 347, in _proto_message
  File "mitmproxy\addons\dumper.py", line 277, in match
  File "mitmproxy\optmanager.py", line 190, in __getattr__
  File "mitmproxy\optmanager.py", line 58, in current
  File "mitmproxy\optmanager.py", line 54, in default
RecursionError: maximum recursion depth exceeded
```

I'm guarding against `tcp_message` getting called again as a result of the injection, so I guess this is happening elsewhere within the library?

**Reply by Prinzhorn** [reply]
Works for me, your guard is likely broken:

Addon:

```py
from mitmproxy import ctx

def tcp_message(flow):
    if flow.messages[-1].content == b"inject\n":
        ctx.master.commands.call("inject.tcp", flow, True, b"I was injected")
```

Terminal 1:

```sh
./mitmdump --scripts tcp_inject.py 
```

Terminal 2:

```sh
ncat -l 1337
```

Terminal 3:

```sh
ncat --proxy localhost:8080 --proxy-type http localhost 1337 
```

![image](https://github.com/mitmproxy/mitmproxy/assets/679144/74a85354-db0c-4f76-a8d3-2917a84d9a9b)
